### PR TITLE
feat (issue #12): add sorting algorithms with benchmarks and comprehensive tests 

### LINKED ISSUE #12: Feature Request — Sorting Algorithms
# Feature Request — Sorting Algorithms

**Summary**
Add four sorting algorithm implementations to `src/sorting/`: **Timsort**, **Merge Sort**, **Quick Sort**, and **Bubble Sort**. Each implementation must be TypeScript-first, well-typed, non-mutating (returns new array), and accept an optional comparator for custom types. This feature aims to provide both educational and production-ready sorting utilities with clear docs and tests.

---

## Goals / Motivation

* Provide a canonical, well-typed set of sorting utilities for the project.
* Offer both educational (Bubble) and high-performance (Timsort, Merge, Quick) implementations.
* Ensure correctness, stability where applicable, and consistent API across algorithms.

---

## Proposed file structure

```
src/
  sorting/
    comparator.ts        # Comparator type + defaultComparator
    bubbleSort.ts
    mergeSort.ts
    quickSort.ts
    timSort.ts
    index.ts

tests/
  sorting/
    bubbleSort.test.ts
    mergeSort.test.ts
    quickSort.test.ts
    timSort.test.ts
```

---

## API & Types

* **Comparator**

```ts
export type Comparator<T> = (a: T, b: T) => number;
export const defaultComparator = <T>(a: any, b: any) => (a === b ? 0 : a < b ? -1 : 1);
```

* Each algorithm signature:

```ts
export function <name><T = number>(arr: T[], cmp?: Comparator<T>): T[]
```

* **Non-mutating**: all functions must return a new array and not change the input.

---

### Bubble Sort

* Purpose: educational reference, O(n²), stable.
* Keep an early-exit optimization (swap flag).
* Use for small arrays in examples.

### Merge Sort

* Purpose: stable, O(n log n) worst-case, predictable memory allocation.
* Implement a top-down recursive version returning a new array.
* Document space complexity and recursion considerations.

### Quick Sort

* Purpose: fast average-case, O(n log n), but O(n²) worst-case.
* Use randomized pivot or median-of-three to reduce worst-case risk.
* Implement in-place helper internally but expose a non-mutating API (work on a copy).
* Document instability and when to prefer Merge/Timsort.

### Timsort

* Purpose: production-grade, stable, best practical performance on real data (runs-based).
* Implement a simplified, correct Timsort variant (merge runs, use insertion sort for small runs, galloping optional).
* If full Timsort is too large, provide a well-documented, high-quality simplified implementation and note differences.

---

## Performance & Type considerations

* Always accept a typed comparator so callers can sort complex objects.
* Document time & space complexity for each algorithm in JSDoc.
* Provide bench notes in README (small examples comparing performance on random and partially-sorted data).
* Keep implementations concise and readable — prefer clarity over clever micro-optimizations.

---

## Tests

Each algorithm must have tests covering:

* empty array
* single element
* already sorted
* reverse sorted
* random arrays (multiple seeds)
* duplicates
* negative numbers / mixed values
* custom comparator (objects)
* immutability (original array unchanged)

Aim for **≥10 test cases per algorithm**. Use Vitest (or existing test runner) consistent with repo.

---

## Documentation

* Update `README.md`  with:

  * short description for each algorithm
  * example usage snippet for numbers and objects (with comparator)
  * complexity table (time/space/stability)
  * note: "Timsort is the recommended default for general use"

---

## Acceptance Criteria

* [ ] `src/sorting/` contains `bubbleSort.ts`, `mergeSort.ts`, `quickSort.ts`, `timSort.ts`, and `comparator.ts`.
* [ ] Each function is fully typed and returns a new array.
* [ ] Tests for each algorithm pass and cover required cases.
* [ ] README updated with usage, examples, and complexity notes.
* [ ] Code passes linting and style checks.
* [ ] Benchmarks: include a simple `bench/sorting` script that compares Merge, Quick, Timsort on random and partially-sorted datasets and commit the results or a short summary.
* [ ] Changelog: add an entry in `CHANGELOG.md` or `RELEASE_NOTES.md` describing the new sorting utilities and recommended default (Timsort).
* [ ] PR metadata: the Pull Request must reference this issue, list implemented algorithms, and assign at least one reviewer.

---

## Implementation Tips

* Reuse `Comparator` across files to keep signatures identical.
* For Quick Sort, implement randomized pivot selection to avoid adversarial arrays.
* For Timsort, keep insertion sort threshold configurable (e.g., 32).
* Keep helper functions `internal` (not exported) and documented.

---

## Suggested PR template text

```
Implemented: <algorithm names>
Files changed: src/sorting/<files>
Notes:
- Algorithm choices and trade-offs
- Test coverage summary
- Bench notes (if any)

Checklist:
- [ ] Non-mutating
- [ ] Tests added
- [ ] README updated
- [ ] Linting passed
```

### PR DESCRIPTION
**Pull request by AntoineDeveloppeur**
# Implement Sorting Algorithms Suite

Fixes #12

## Summary

Implemented four sorting algorithms with a flexible comparator system, comprehensive test coverage, and performance benchmarking suite.

## Algorithms Implemented

- **Bubble Sort** - O(n²) average case, stable, educational purposes
- **Merge Sort** - O(n log n) guaranteed, stable, predictable performance  
- **Quick Sort** - O(n log n) average case with median-of-three pivot selection
- **Tim Sort** - Simplified implementation, O(n log n), optimized for real-world data

### Core Implementation

- `src/sorting/bubbleSort.ts` - Bubble sort with early-exit optimization
- `src/sorting/mergeSort.ts` - Pure merge sort implementation
- `src/sorting/quickSort.ts` - Quick sort with median-of-three pivot strategy
- `src/sorting/timSort.ts` - Simplified Tim sort (no galloping mode)
- `src/sorting/comparator.ts` - Default comparator for primitive types
- `src/index.ts` - Export sorting functions from main entry point

### Testing & Benchmarking

- `tests/sorting/*.test.ts` - Comprehensive unit tests for all algorithms
- `bench/sorting/benchmark.ts` - Performance benchmarking framework
- `bench/sorting/utils.ts` - Benchmark-specific utilities (array generators, timing)
- `bench/sorting/results.md` - Detailed performance analysis and recommendations

## Implementation Notes

I had to make few choices along the way. There are all mentionned below. Please tell me if I need to change something.

### Algorithm Design Choices

**Non-Mutating Implementations**: All sorting functions return new arrays without modifying the input, ensuring functional purity and preventing side effects.

**Generic Comparator System**: Implemented function overloading to enforce type safety:

- Default comparator handles `number | string | bigint` primitives
- Custom comparators required for complex types (enforced at compile-time)
- Prevents runtime errors from incorrect type assumptions

**Quick Sort Pivot Strategy**: Uses median-of-three pivot selection (first, middle, last elements) instead of randomized pivot. This approach provides better performance on partially sorted data while avoiding the overhead of random number generation.

**Tim Sort Simplification**: Implemented a simplified version without galloping mode. I tried a more complex one by benchmark showed that galloping and others options overhead degraded performance compared to the streamlined approach. I tried this one : [https://github.com/Torathion/timsort2/blob/master/src/index.ts#L77](https://github.com/Torathion/timsort2/blob/master/src/index.ts#L77)

The current implementation maintains Tim Sort's core advantages (hybrid insertion/merge strategy). 

### Performance Findings

Benchmark results reveal important performance characteristics:

**JavaScript Native Sort Dominance (Small Arrays)**:

- Native `Array.prototype.sort()` is 2-10x faster for arrays < 4,000 elements
- V8 engine optimizations make it the best choice for typical web application use cases

**Tim Sort Excellence (Large Arrays)**:

- Best performer for arrays > 10,000 elements
- 50,000 random elements: **16.6ms** (2.2x faster than native sort)
- 50,000 partially sorted: **11.4ms** (31% faster than random data)
- Scales exceptionally well as dataset size increases

**Recommendation**: Use native sort for small/medium arrays; prefer Tim Sort for large datasets or when data may be partially sorted. 
This recommendation is not written in the readme. Please confirm for me to change readme.

Full benchmark analysis available in `bench/sorting/results.md`.

### File Structure Decisions

**Main Export Location**: `src/index.ts` rather than creating `src/sorting/index.ts`. This maintains consistency with the existing project structure where all public APIs are exported from the root index.

**Benchmark Utilities Placement**: Created benchmark-specific utilities in `bench/sorting/utils.ts` instead of `src/utils/`. These utilities (`ArrayGenerator`, `RandomInt`, timing functions) are only used for performance testing and don't belong in the production source code.

## Checklist

- [x] **Non-mutating** - All functions return new arrays without modifying inputs
- [x] **Tests added** - Comprehensive test coverage for all algorithms and edge cases
- [x] **README updated** - Documentation includes usage examples and performance recommendations  
- [x] **Linting passed** - Code follows project style guidelines
- [x] **Benchmark suite** - Run with `npm run benchmark:sorting`
- [x] **Performance analysis** - Detailed results in `bench/sorting/results.md`

## Running Benchmarks

```bash
npm run benchmark:sorting
```

## Summary by Sourcery

Add a non-mutating, comparator-based sorting suite with benchmarking and documentation, and expose the new algorithms from the public API.

New Features:
- Introduce generic comparator type utilities for type-safe custom comparison functions.
- Add bubble sort, merge sort, quick sort, and TimSort-based implementations that work with built-in and custom comparator functions.
- Expose the new sorting algorithms from the main library entry point for public consumption.
- Provide a benchmark harness and CLI entry to compare sorting algorithm performance, including against JavaScript's native sort.
- Document the sorting APIs, usage patterns, and relative algorithm characteristics in the README.

Enhancements:
- Add a changelog entry for version 1.2.0 describing the new sorting capabilities and comparator support.

Build:
- Add an npm script to run the sorting benchmarks via the TSX-based benchmark entrypoint.

Documentation:
- Extend README with a sorting section, including comparator examples, per-algorithm descriptions, and a complexity comparison table.
- Add a benchmark results document summarizing performance findings and recommendations for choosing algorithms by data size and shape.

Tests:
- Add unit tests for sorting algorithms, comparator behavior, benchmark utilities, and array generation helpers to ensure correctness and benchmark integrity.
- Add tests around the Benchmark orchestration to validate table structure, array generation, and timing output formatting.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -213,6 +211,165 @@ flatten([1, 2, 3]); // [1, 2, 3]
...
+const younger: Person = { name: "Maria", age: 18 };
+const older: Person = { name: "John", age: 23 };
+compareByAge(younger, older); // -1 (Maria comes first)
+compareByAge(older, younger); // 1 (John comes after)
+compareByAge(younger, younger); // 0 (equal)
+```
+
+</details>
+
+The fastest is TimSort
```
[the remark this answers is not in the document]
Comment (AntoineDeveloppeur) [reply README.md]: @sourcery-ai resolve

**Comment by AntoineDeveloppeur** [comment]
Hi @parthik-jadav , PR is ready for review. Bot raised issues are resolved.

**Comment by AntoineDeveloppeur** [comment]
Hi @parthikjadav, could you review PR ?

**Comment by parthikjadav** [comment]
can you please solve the isssues suggest by the sourcery, btw thanks for the contribution

**Comment by AntoineDeveloppeur** [comment]
@parthikjadav I did it already. All checks have passed on my screen. It is written "sourcery review : completed successfully"
Capture d’écran 2026-02-17 082554
